### PR TITLE
Bug 2000607: Add an alert for the domain mapping linked with other knative service

### PR DESCRIPTION
**Pull request by karthikjeeyar**
**Fixes**: 

https://issues.redhat.com/browse/ODC-6305

**Analysis / Root cause**: 

Moving from the domain from one resource to another is not intuitive. The user should be given a warning or message.

**Solution Description**: 

1. Add an alert in the route section if  any one of the selected  domain mappings is already linked to a different knative service 
2. Add the knative service name next to the domain mapping as a dropdown item.
3. Added a extra validation so that newly created name should contain two segments seperated by dot.

**Screen shots / Gifs for design review**: 

GIF:

mp-domain

Alert for using a domain mapping from different knative service:
image
Dropdown name with ksvc name :
image

**Unit test coverage report**: 

**Serverless-utils:**
```
  hasOtherKsvcDomainMappings
    ✓ should return false if the domain mapping is empty
    ✓ should return false if the other ksvc info is not present in domain mapping
    ✓ should return true if the other ksvc info is present in domain mapping
  removeDuplicateDomainMappings
    ✓ should return empty arary if the domain mapping empty or null (1ms)
    ✓ should return unique value in the array (1ms)
    ✓ should strip the ksvc info and return unique value in the array
  getOtherKsvcFromDomainMapping
    ✓ should return null for invalid values
    ✓ should return null the ksvc name if is matching with the current ksvc name
    ✓ should return the ksvc name if its not matching with the current ksvc name (1ms)
  removeKsvcInfoFromDomainMapping
    ✓ should return empty string
    ✓ should return the same domain name if the ksvc info is not in the string
    ✓ should return only the domain name if the ksvc info is the string
   getAllOtherDomainMappingInUse
    ✓ should return empty array for the empty input
    ✓ should return empty array if selected domain is not already in the cluster (1ms)
    ✓ should return empty if the selected domain is mapped to current knative service
    ✓ should return domains if the selected domains are not mapped to current Knative service (1ms)
     ✓ should not return domains if the user creates domain with the same name again
```
**ServerlessRouteSection.spec.tsx:**
```
    ✓ Should render SelectInputField with the domain mapping options without other ksvc info (3ms)
    ✓ Should not contain the warning Alert if the selected domain is not from other knative services (3ms)
    ✓ Should contain the warning Alert if any of the selected domains is from other knative services (2ms)
```

**Test setup:**

1. Create a serverless workload with domain mapping
2. Check the domain mapping details in the workload sidebar
3. Create another workload using same domain mapping
4. Check the domain mapping details on the first workload

**Browser conformance**: 

- [x] Chrome
- [x] Firefox
- [ ] Safari
- [ ] Edge

cc: @invincibleJai

### CODE REVIEW

File: frontend/packages/dev-console/src/components/import/serverless/ServerlessRouteSection.tsx [hunk 1]
```diff
@@ -72,16 +84,30 @@ const ServerlessRouteSection: React.FC = () => {
...
-          label={t('devconsole~Domain mapping')}
-          options={domainMappingResources}
-          placeholderText={t('devconsole~Add domain')}
-          helpText={t('devconsole~Mapping a custom domain to the Knative service')}
-          isCreatable
-          hasOnCreateOption
-        />
+        <>
+          <SelectInputField
+            data-test-id="kafkasource-bootstrapservers-field"
```
Comment (vikram-raj) [inline frontend/packages/dev-console/src/components/import/serverless/ServerlessRouteSection.tsx:89]: nit: Can you please pdate this `data-test-id` as it is not related to this component.
```suggestion
            data-test-id="domain-mapping-field"
```

File: frontend/packages/dev-console/src/components/import/serverless/ServerlessRouteSection.tsx [hunk 2]
```diff
@@ -72,16 +84,30 @@ const ServerlessRouteSection: React.FC = () => {
...
+          />
+          {hasOtherKsvcDomainMappings(serverless.domainMapping) && (
+            <Alert
+              data-test="domain-mapping-warning"
+              variant="warning"
+              isInline
+              title={t('devconsole~Domain mapping(s) will be updated')}
+            >
+              {t(
+                'devconsole~Some of the selected domain mappings already reference Serverless deployments in this namespace.  These Serverless deployments will be updated to no longer map to the domain mapping.',
```
Comment (vikram-raj) [inline frontend/packages/dev-console/src/components/import/serverless/ServerlessRouteSection.tsx:106]:
```suggestion
                'devconsole~Some of the selected domain mappings already reference Serverless deployments in this namespace. These Serverless deployments will be updated to no longer map to the domain mapping.',
```

File: frontend/packages/dev-console/src/components/import/serverless/ServerlessRouteSection.tsx [hunk 3]
```diff
@@ -72,16 +84,30 @@ const ServerlessRouteSection: React.FC = () => {
...
-          hasOnCreateOption
-        />
+        <>
+          <SelectInputField
+            data-test-id="domain-mapping-field"
+            name="serverless.domainMapping"
+            label={t('devconsole~Domain mapping')}
+            options={domainMappingResources}
+            placeholderText={t('devconsole~Add domain')}
+            helpText={t('devconsole~Enter custom domain to map to the Serverless Deployment')}
```
Comment (vikram-raj) [inline frontend/packages/dev-console/src/components/import/serverless/ServerlessRouteSection.tsx:94]:
```suggestion
            helpText={t('devconsole~Enter custom domain to map to the Serverless deployment')}
```

Comment (karthikjeeyar) [reply frontend/packages/dev-console/src/components/import/serverless/ServerlessRouteSection.tsx]: Updated the text based on the discussion over the tmp-odc-serverless slack channel.

**Comment by vikram-raj** [comment]
Seeing the same bullets point twice after adding the same domain name.

image

**Comment by karthikjeeyar** [comment]
@vikram-raj Fixed the  user created duplicate entry in the warning message and added the tests for the same.

File: frontend/packages/dev-console/src/components/import/serverless/__tests__/serverless-utils.spec.ts [hunk 4]
```diff
@@ -50,3 +58,111 @@ describe('serverless-utils', () => {
...
+    expect(hasOtherKsvcDomainMappings(['abc.org', 'example.org', 'domain.ksvc1 (ksvc1)'])).toBe(
+      true,
+    );
+  });
+});
+
+describe('removeDuplicateDomainMappings', () => {
+  it('should return empty arary if the domain mapping empty or null', () => {
+    expect(removeDuplicateDomainMappings(null, null)).toEqual([]);
+    expect(removeDuplicateDomainMappings(['a'], null)).toEqual([]);
```
Comment (invincibleJai) [inline frontend/packages/dev-console/src/components/import/serverless/__tests__/serverless-utils.spec.ts:81]: ideally, this should return `['a']`. WDYT?

File: frontend/packages/dev-console/src/components/import/serverless/__tests__/serverless-utils.spec.ts [hunk 5]
```diff
@@ -50,3 +58,111 @@ describe('serverless-utils', () => {
...
+    ).toEqual(['domain.org']);
+  });
+});
+
+describe('getOtherKsvcFromDomainMapping', () => {
+  it('should return null for invalid values', () => {
+    expect(getOtherKsvcFromDomainMapping({}, '')).toBeNull();
+  });
+
+  it('should return null the ksvc name if is matching with the current ksvc name', () => {
```
Comment (invincibleJai) [inline frontend/packages/dev-console/src/components/import/serverless/__tests__/serverless-utils.spec.ts:101]: nit: 
```suggestion
  it('should return null as the ksvc name if it is matching with the current ksvc name', () => {
```

File: frontend/packages/dev-console/src/components/import/serverless/__tests__/serverless-utils.spec.ts (hunk 4 above)
Comment (karthikjeeyar) [reply frontend/packages/dev-console/src/components/import/serverless/__tests__/serverless-utils.spec.ts]: added a additional test for it.

**Comment by vikram-raj** [comment]
I can enter the same domain name twice. This should not happen.

image

**Review by vikram-raj — COMMENTED** [review]
Verified it and it works as expected. 
/lgtm

**Review by serenamarie125 — APPROVED** [review]
LGTM!
Tx for taking care of this

**Pull request merged**